File: AppRemover/src/main/java/com/vs_unusedappremover/data/ApplicationCollection.java
package com.vs_unusedappremover.data;

import android.content.ContentValues;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.database.Cursor;
import android.database.DataSetObserver;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.google.common.base.Function;
import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.vs_unusedappremover.AppEntry;
import com.vs_unusedappremover.MyApplication;
import com.vs_unusedappremover.PackageSize;
import com.vs_unusedappremover.data.DatabaseHelper.AppTable;

import java.io.File;
import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

import static java.util.Collections.singletonList;

public class ApplicationCollection {
	
	private static final String TAG = ApplicationCollection.class.getSimpleName();
	private static final long NOTIFY_UPDATED_INTERVAL = 500;
	
	private HashMap<String, AppEntry> previousData = new HashMap<>();
	private final Context context;
	private final HashSet<DataSetObserver> observers = new HashSet<>();
	private final GroupChangeNotifications groupedChangeNotifications = new GroupChangeNotifications(NOTIFY_UPDATED_INTERVAL);
	private final MyApplication app = MyApplication.getInstance();
	private final Object modificationLock = new Object();
	
	public ApplicationCollection(Context context) {
		this.context = context;
		notifyChanged();
	}
	
	public ArrayList<AppEntry> values(Predicate<AppEntry> filter) {
		return values(filter, null);
	}
	
	public ArrayList<AppEntry> values(Predicate<AppEntry> filter, Comparator<AppEntry> order) {		
		synchronized (modificationLock) {
			ArrayList<AppEntry> result = Lists.newArrayList(Collections2.filter(getData().values(), filter));
			if (order != null) {
				Collections.sort(result, order);
			}
			return result;
		}
	}
			
	public void notifyUsed(String appPackage, long when, AppEntry.RanIn ranIn) {
		long now = System.currentTimeMillis();
		synchronized (modificationLock) {
			AppEntry entry = getData().get(appPackage);
			if (entry != null && entry.lastUsedTime < when) {				
				entry.lastUsedTime = when;
                entry.ranIn = ranIn;
				groupedChangeNotifications.add(entry);
				Log.i(TAG, appPackage + " used " + (now - when) / 1000 + "sec ago");
			}			
		}
	}
	
	public void setNotifyAbout(String appPackage, boolean value) {
		synchronized (modificationLock) {
			AppEntry entry = getData().get(appPackage);
			if (entry != null && entry.notifyAbout != value) {
				entry.notifyAbout = value;
				saveToCache(singletonList(entry));
				fireChanged();
			}			
		}
	}
	
	public void addObserver(DataSetObserver o) {
		if (o != null) {
			observers.add(o);
		}
	}
	
	public void removeObserver(DataSetObserver o) {
		if (o != null) {
			observers.remove(o);
		}
	}
	
	public void notifyChanged() {
		final PackageManager pm = context.getPackageManager();

		synchronized (modificationLock) {

			HashMap<String, AppEntry> data = new HashMap<String, AppEntry>();
			for (ApplicationInfo appInfo : pm.getInstalledApplications(0)) {
				if (Applications.isThirdParty(appInfo)) {
				
					AppEntry entry = previousData.get(appInfo.packageName);
					if (entry == null) {					 
						entry = new AppEntry();
						entry.info = appInfo;
					    entry.size = PackageSize.UNKNOWN;
                        entry.ranIn = AppEntry.RanIn.UNKNOWN;
                        entry.label = appInfo.packageName;
					    fillFromDbCache(entry);			    
					}
					data.put(appInfo.packageName, entry);
					app.getBackgroundExecutor().execute(new LoadExpensiveInfo(entry));
				}
			}
			previousData = data;
		}
		
		fireChanged();
	}

	private HashMap<String, AppEntry> getData() {
		synchronized (modificationLock) {
			return previousData;
		}
	}
	
	private void fillFromDbCache(AppEntry entry) {
		DatabaseHelper dbHelper = app.getDbHelper();
		
		Cursor c = new SelectQuery()
			.select(AppTable.TIME_LAST_USED, AppTable.TIME_INSTALLED, 
					AppTable.DOWLOAD_COUNT, AppTable.RATING, 
					AppTable.SIZE, AppTable.NOTIFY, AppTable.RAN_IN)
			.from(AppTable.NAME)
			.where(AppTable.PACKAGE + "=?", entry.info.packageName)
			.execute(dbHelper.getReadableDatabase());
				
		try {
			if (c.moveToFirst()) {
				entry.lastUsedTime = c.getLong(0);
                entry.ranIn = AppEntry.RanIn.byId(c.getInt(6));
				entry.installTime = c.getLong(1);
				entry.downloadCount = c.getString(2);
				entry.rating = c.getFloat(3);
				entry.size = c.getLong(4);
				entry.notifyAbout = (c.getInt(5) != 0);
			}
		} finally {
			c.close();
		}
	}

	private void saveToCache(Iterable<AppEntry> entries) {		
		Set<String> existingApps = getExistingApps();
		
		SQLiteDatabase db = app.getDbHelper().getWritableDatabase();
		db.beginTransaction();
		try {			
			ContentValues values = new ContentValues();
			
			for (AppEntry entry : entries) {
											 
				values.put(AppTable.TIME_INSTALLED, entry.installTime);
				values.put(AppTable.TIME_LAST_USED, entry.lastUsedTime);
				values.put(AppTable.DOWLOAD_COUNT, entry.downloadCount);
				values.put(AppTable.RATING, entry.rating);
				values.put(AppTable.SIZE, entry.size);
				values.put(AppTable.NOTIFY, entry.notifyAbout);
                values.put(AppTable.RAN_IN, entry.ranIn.id);

                if (existingApps.contains(entry.info.packageName)) {
					db.update(AppTable.NAME, values, AppTable.PACKAGE + "=?", new String[] { entry.info.packageName });
				} else {
					values.put(AppTable.PACKAGE, entry.info.packageName);
					db.insertWithOnConflict(AppTable.NAME, null, values, SQLiteDatabase.CONFLICT_REPLACE);
				}
				values.clear();
			}
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
				
		final DateFormat time = DateFormat.getTimeInstance(DateFormat.SHORT);
		Log.i(TAG, " cached " + Iterables.toString(Iterables.transform(entries, new Function<AppEntry, String>() {
			@Override 
			public String apply(AppEntry e) {
				return e.info.packageName + "=(" +  time.format(new Date(e.lastUsedTime)) + ")";
			}
		})));		
	}
	
	private Set<String> getExistingApps() {
		SQLiteDatabase db = app.getDbHelper().getReadableDatabase();
		Cursor c = new SelectQuery()
			.select(AppTable.PACKAGE)
			.from(AppTable.NAME)
			.execute(db);
		
		HashSet<String> packages = new HashSet<String>(c.getColumnCount());
		try {
			while (c.moveToNext()) {
				packages.add(c.getString(0));
			}
			return packages;
		} finally {
			c.close();
		}			
	}
	
	private void fireChanged() {
		for (DataSetObserver o : observers) {
			o.onChanged();
		}
	}
		
	private class LoadExpensiveInfo implements Runnable {

		private final AppEntry entryToFill;
		private final PackageManager pm = context.getPackageManager();
		
		public LoadExpensiveInfo(AppEntry entry) {
			this.entryToFill = entry;
		}
		
		@Override
		public void run() {			
			File apk = new File(entryToFill.info.sourceDir);
			long size = PackageSize.get(pm, entryToFill.info);
			String packageName = entryToFill.info.packageName;
			String label = apk.exists() ? entryToFill.info.loadLabel(pm).toString() : packageName;
			long installTime = (entryToFill.installTime == 0) ? getFirstInstallTime(packageName) : entryToFill.installTime;
			
			if (size != entryToFill.size ||
				!label.equals(entryToFill.label) ||
				installTime != entryToFill.installTime) 
			{
				synchronized (modificationLock) {
					entryToFill.size = size;
					entryToFill.label = label;
					entryToFill.installTime = installTime;
				}
				groupedChangeNotifications.add(entryToFill);
			}
		}

		private long getFirstInstallTime(String packageName) {
			try {
				return pm.getPackageInfo(packageName, 0).firstInstallTime;
			} catch (NameNotFoundException e) {
				return 0;
			}
		}
	}

	private class GroupChangeNotifications implements Runnable {
		
		private final long updatePeriodMillis;
		private final HashMap<String, AppEntry> changes = new HashMap<String, AppEntry>();
		
		public GroupChangeNotifications(long updatePeriodMillis) {
			this.updatePeriodMillis = updatePeriodMillis;
		}
		
		public synchronized void add(AppEntry entry) {
			boolean wasEmpty = (changes.size() == 0);
			changes.put(entry.info.packageName, entry);
			if (wasEmpty) {
				app.getUiHandler().postDelayed(this, updatePeriodMillis);
			}
		}
		
		public synchronized int flush() {
			final ArrayList<AppEntry> changedList = new ArrayList<AppEntry>(changes.values());
			app.getBackgroundExecutor().execute(new Runnable() {				
				@Override
				public void run() {
					saveToCache(changedList);					
				}
			});			
			int size = changes.size();
			changes.clear();
			return size;
		}
		
		@Override
		public void run() {
			int count = flush();
			Log.i(TAG, "flushed " + count + " entries to cache");
			fireChanged();
		}
	}
}

File: AppRemover/src/main/java/com/vs_unusedappremover/data/ApplicationCollection.kt
package com.vs_unusedappremover.data

import android.content.ContentValues
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager.NameNotFoundException
import android.database.DataSetObserver
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import com.vs_unusedappremover.AppEntry
import com.vs_unusedappremover.MyApplication
import com.vs_unusedappremover.PackageSize
import com.vs_unusedappremover.data.DatabaseHelper.AppTable
import java.io.File
import java.text.DateFormat
import java.util.*
import kotlin.collections.ArrayList

class ApplicationCollection(private val context: Context) {

    private var previousData = HashMap<String, AppEntry>()
    private val observers = HashSet<DataSetObserver>()
    private val groupedChangeNotifications = GroupChangeNotifications(NOTIFY_UPDATED_INTERVAL)
    private val app = MyApplication.instance
    private val modificationLock = Any()
    private val data: HashMap<String, AppEntry>
        get() {
            synchronized(modificationLock) {
                return previousData
            }
        }

    init {
        notifyChanged()
    }

    @JvmOverloads
    fun values(order: Comparator<AppEntry>? = null, filter: (AppEntry) -> Boolean): ArrayList<AppEntry> {
        synchronized(modificationLock) {
            val result = data.values.filter(filter)
            if (order != null) {
                Collections.sort(result, order)
            }
            return ArrayList(result)
        }
    }

    fun notifyUsed(appPackage: String, `when`: Long, ranIn: AppEntry.RanIn) {
        val now = System.currentTimeMillis()
        synchronized(modificationLock) {
            val entry = data[appPackage]
            if (entry != null && entry.lastUsedTime < `when`) {
                entry.lastUsedTime = `when`
                entry.ranIn = ranIn
                groupedChangeNotifications.add(entry)
                Log.i(TAG, appPackage + " used " + (now - `when`) / 1000 + "sec ago")
            }
        }
    }

    fun setNotifyAbout(appPackage: String, value: Boolean) {
        synchronized(modificationLock) {
            val entry = data[appPackage]
            if (entry != null && entry.notifyAbout != value) {
                entry.notifyAbout = value
                saveToCache(listOf<AppEntry>(entry))
                fireChanged()
            }
        }
    }

    fun addObserver(o: DataSetObserver?) {
        if (o != null) {
            observers.add(o)
        }
    }

    fun removeObserver(o: DataSetObserver?) {
        if (o != null) {
            observers.remove(o)
        }
    }

    fun notifyChanged() {
        val pm = context.packageManager

        synchronized(modificationLock) {

            val data = HashMap<String, AppEntry>()
            for (appInfo in pm.getInstalledApplications(0)) {
                if (appInfo.isThirdParty()) {
                    var entry = previousData[appInfo.packageName]
                    if (entry == null) {
                        entry = AppEntry(info = appInfo,
                                label = appInfo.packageName,
                                size = PackageSize.UNKNOWN,
                                ranIn = AppEntry.RanIn.UNKNOWN)
                        fillFromDbCache(entry)
                    }
                    data.put(entry.label, entry)
                    app.backgroundExecutor.execute(LoadExpensiveInfo(entry))
                }
            }
            previousData = data
        }

        fireChanged()
    }

    private fun ApplicationInfo.isThirdParty(): Boolean {
        if (this.flags and ApplicationInfo.FLAG_UPDATED_SYSTEM_APP != 0) return false
        return this.flags and ApplicationInfo.FLAG_SYSTEM == 0
    }

    private fun fillFromDbCache(entry: AppEntry) {
        SelectQuery()
                .select(AppTable.TIME_LAST_USED, AppTable.TIME_INSTALLED,
                        AppTable.DOWNLOAD_COUNT, AppTable.RATING,
                        AppTable.SIZE, AppTable.NOTIFY, AppTable.RAN_IN)
                .from(AppTable.NAME)
                .where(AppTable.PACKAGE + "=?", entry.info.packageName)
                .execute(app.dbHelper.readableDatabase).use {
            if (it.moveToFirst()) {
                entry.apply {
                    lastUsedTime = it.getLong(0)
                    ranIn = AppEntry.RanIn.byId(it.getInt(6))
                    installTime = it.getLong(1)
                    downloadCount = it.getString(2)
                    rating = it.getFloat(3)
                    size = it.getLong(4)
                    notifyAbout = it.getInt(5) != 0
                }
            }
        }
    }

    private fun saveToCache(entries: Iterable<AppEntry>) {
        val existingApps = existingApps

        app.dbHelper.writableDatabase.inTransaction {
            val values = ContentValues()
            for (entry in entries) {
                values.apply {
                    put(AppTable.TIME_INSTALLED, entry.installTime)
                    put(AppTable.TIME_LAST_USED, entry.lastUsedTime)
                    put(AppTable.DOWNLOAD_COUNT, entry.downloadCount)
                    put(AppTable.RATING, entry.rating)
                    put(AppTable.SIZE, entry.size)
                    put(AppTable.NOTIFY, entry.notifyAbout)
                    put(AppTable.RAN_IN, entry.ranIn.id)

                    if (existingApps.contains(entry.info.packageName)) {
                        update(AppTable.NAME, this, AppTable.PACKAGE + "=?", arrayOf(entry.info.packageName))
                    } else {
                        put(AppTable.PACKAGE, entry.info.packageName)
                        insertWithOnConflict(AppTable.NAME, null, this, SQLiteDatabase.CONFLICT_REPLACE)
                    }
                    clear()
                }
            }
        }

        val time = DateFormat.getTimeInstance(DateFormat.SHORT)

        Log.i(TAG, " cached " + entries.joinToString(prefix = "[", postfix = "]") { it ->
            it.info.packageName + "=(" + time.format(Date(it.lastUsedTime)) + ")"
        })
    }

    private inline fun SQLiteDatabase.inTransaction(func: SQLiteDatabase.() -> Unit) {
        beginTransaction()
        try {
            func()
            setTransactionSuccessful()
        } finally {
            endTransaction()
        }
    }

    private val existingApps: Set<String>
        get() {
            SelectQuery()
                    .select(AppTable.PACKAGE)
                    .from(AppTable.NAME)
                    .execute(app.dbHelper.readableDatabase)
                    .use {
                        val packages = HashSet<String>(it.columnCount)
                        while (it.moveToNext()) {
                            packages.add(it.getString(0))
                        }
                        return packages
                    }
        }

    private fun fireChanged() {
        for (o in observers) {
            o.onChanged()
        }
    }

    private inner class LoadExpensiveInfo(private val entryToFill: AppEntry) : Runnable {
        private val pm = context.packageManager

        override fun run() {
            val apk = File(entryToFill.info.sourceDir)
            val size = PackageSize.get(pm, entryToFill.info)
            val packageName = entryToFill.info.packageName
            val label = if (apk.exists()) entryToFill.info.loadLabel(pm).toString() else packageName
            val installTime = if (entryToFill.installTime == 0L) getFirstInstallTime(packageName) else entryToFill.installTime

            if (size != entryToFill.size ||
                    label != entryToFill.label ||
                    installTime != entryToFill.installTime) {
                synchronized(modificationLock) {
                    entryToFill.size = size
                    entryToFill.label = label
                    entryToFill.installTime = installTime
                }
                groupedChangeNotifications.add(entryToFill)
            }
        }

        private fun getFirstInstallTime(packageName: String): Long {
            try {
                return pm.getPackageInfo(packageName, 0).firstInstallTime
            } catch (e: NameNotFoundException) {
                return 0
            }

        }
    }

    private inner class GroupChangeNotifications(private val updatePeriodMillis: Long) : Runnable {
        private val changes = HashMap<String, AppEntry>()

        @Synchronized
        fun add(entry: AppEntry) {
            val wasEmpty = changes.size == 0
            changes.put(entry.info.packageName, entry)
            if (wasEmpty) {
                app.uiHandler.postDelayed(this, updatePeriodMillis)
            }
        }

        @Synchronized
        fun flush(): Int {
            val changedList = ArrayList(changes.values)
            app.backgroundExecutor.execute { saveToCache(changedList) }
            val size = changes.size
            changes.clear()
            return size
        }

        override fun run() {
            val count = flush()
            Log.i(TAG, "flushed $count entries to cache")
            fireChanged()
        }
    }

    companion object {

        private val TAG = ApplicationCollection::class.java.simpleName
        private val NOTIFY_UPDATED_INTERVAL: Long = 500
    }
}
